dags/iop/parser.py

```diff
@@ -106,6 +117,51 @@ def _get_extracted_arxiv_eprint_value(self, article):
         except AttributeError:
             self.logger.error("No arXiv eprints found", dois=self.dois)
 
+    def _get_date_element(self, article):
+        try:
+            date = (
+                article.find("front/article-meta/history/date[@date-type='published']")
+                or article.find("front/article-meta/history/date[@date-type='epub']")
+                or article.find("front/article-meta/history/pub-date[@pub-type='ppub']")
+                or article.find("front/article-meta/pub-date")
+            )
+            return date
+        except AttributeError:
+            return
+
+    def _get_date_published(self, article):
+        date_element = self._get_date_element(article)
+        return self._get_date(date_element)
+
+    def _get_date(self, date):
+        try:
+            year = int(date.find("year").text)
+            self.year = year
+        except NODE_ATTRIBUTE_NOT_FOUND_ERRORS:
+            self.logger.error("Cannot find year of date_published in XML")
+        try:
+            month = int(date.find("month").text)
+        # was discussed with Anne: if the day is not present, we return year and month
+        # if the month value is not present - we return just year
+        # if the year is not present - the parsing should crash, because according schema
+        # the publication_year is required, which is take from date_published
+        except NODE_ATTRIBUTE_NOT_FOUND_ERRORS:
+            self.logger.error("Cannot find month of date_published in XML")
+            month = None
+        try:
+            day = int(date.find("day").text)
+        except NODE_ATTRIBUTE_NOT_FOUND_ERRORS:
+            self.logger.error("Cannot find day of date_published in XML")
+            day = None
+        try:
+            date = PartialDate(year=self.year, month=month, day=day)
+            return PartialDate.dumps(date)
+        except NODE_ATTRIBUTE_NOT_FOUND_ERRORS + (ValueError,):
+            return str(self.year)
+
+    def _get_journal_year(self, _):
```

Comment:
why you need to pass `_` arg to the function, is that because the extractor requires it?

Reply:
yes, during the extraction the article object is passed to every extractor: 
https://github.com/cern-sis/workflows/blob/401947d22b5135a5dadbdefdb05bdfe5a7d8eb57/dags/common/parsing/parser.py#L41